src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs

```diff
@@ -253,73 +251,93 @@ private void RemoveUnneededReferences()
             ///  </summary>
             internal void ComputeLambdaScopesAndFrameCaptures()
             {
+                // We need to keep this around
+                Old_RemoveUnneededReferences();
+                RemoveUnneededReferences();
+
                 LambdaScopes = new Dictionary<MethodSymbol, BoundNode>(ReferenceEqualityComparer.Instance);
                 NeedsParentFrame = new HashSet<BoundNode>();
 
-                RemoveUnneededReferences();
+                VisitClosures(_scopeTree, (scope, closure) =>
+                {
+                    if (closure.CapturedVariables.Count > 0)
+                    {
+                        (Scope innermost, Scope outermost) = FindLambdaScopeRange(closure, scope);
+                        RecordClosureScope(innermost, outermost, closure);
+                    }
+                });
 
-                foreach (var kvp in CapturedVariablesByLambda)
+                (Scope innermost, Scope outermost) FindLambdaScopeRange(Closure closure, Scope closureScope)
                 {
-                    var lambda = kvp.Key;
-                    var capturedVars = kvp.Value;
+                    Scope innermost = null;
+                    Scope outermost = null;
 
-                    var allCapturedVars = ArrayBuilder<Symbol>.GetInstance(capturedVars.Count);
-                    allCapturedVars.AddRange(capturedVars);
+                    var capturedVars = PooledHashSet<Symbol>.GetInstance();
+                    capturedVars.AddAll(closure.CapturedVariables);
 
                     // If any of the captured variables are local functions we'll need
                     // to add the captured variables of that local function to the current
                     // set. This has the effect of ensuring that if the local function
                     // captures anything "above" the current scope then parent frame
                     // is itself captured (so that the current lambda can call that
                     // local function).
-                    foreach (var captured in capturedVars)
+                    foreach (var captured in closure.CapturedVariables)
```

Comment:
Specifically, here's my argument: this code finds the min and max scope of all captured variables in the set. Each captured variable is independent and the Min and Max operations for scopes are associative and commutative.